### ISSUE TITLE
Add unit-tests for all the Gambit predefined procedures

### ISSUE THREAD

**Issue by feeley** [gambit/gambit#756]
This issue has a bounty of $100 CAD.

Many of the Gambit predefined procedures have extensive unit tests in the `tests/unit-tests` directory.  For example each file in the `tests/unit-tests/04-list` directory contains unit tests for a specific list processing procedure or set of related procedures, and `tests/unit-tests/09-io` is for I/O related procedures.

These don't cover all the predefined procedures.  For example `tests/unit-tests/09-io` does not contain unit tests for `print`, `println` and `read-line` (which are rather useful and frequently used predefined procedures).

There needs to be a way to list all the predefined procedures that are tested by the unit tests so that this list can be compared with `lib/gambit/gambit.sld` which has a list of all the procedures exported by Gambit (that are not in the `##` namespace).  This could be done by extending the definitions of the unit testing macros in `tests/unit-tests/#.scm` which are used by all the unit tests.

Unit tests need to be added for each predefined procedure not yet tested.

**Comment by tzh476** [comment]
Hi,

I've been working through the coverage analysis this issue asks for (which predefined procedures are not yet tested), using the exported symbols in `lib/gambit/gambit.sld` (1,532 symbols) against the full `tests/unit-tests/` tree (391 files), doing exact token matching on comment-stripped sources.

Summary of what I found:

- 354 exported symbols never appear anywhere in the test corpus.
- 247 of those 354 are already marked `;;unimplemented;;` in the test tree (360 such markers overall), so they are not missing tests — Gambit does not implement them yet, and writing tests for them would be out of scope here.
- That leaves a small set of implemented-but-untested procedures. The cleanest group is the 8 R7RS `bytevector` names exported at `gambit.sld` lines 274-281 (`bytevector`, `bytevector-append`, `bytevector-copy`, `bytevector-copy!`, `bytevector-length`, `bytevector-u8-ref`, `bytevector-u8-set!`, `bytevector?`), plus `make-bytevector` at line 324: the existing tests only cover the `u8vector` spellings, and none of these 9 names appear (unmarked) anywhere in the test sources.

I have a candidate `tests/unit-tests/07-vector/bytevector.scm` written in the direct-assertion style of the existing `07-vector/u8vector.scm` — 100 assertions covering normal, boundary, and error cases (distinguishing type-exception from range-exception from wrong-number-of-arguments-exception), and not touching the groups covered by the existing open PRs.

Verified locally against a `master` build, running `run-unit-tests.scm` from within `tests/`:

- with the file: `FAILED 2 UNIT TESTS OUT OF 372`
- without it: `FAILED 2 UNIT TESTS OUT OF 371`

so the failure count is unchanged and the total drops by one — the new file passes, and the two pre-existing failures (`02-flonum/fladdmul.scm`, `03-number/lcm.scm`) are unrelated to it. It also compiles and runs cleanly under `gsc -exe`.

One note on method, in case it matters for review: unlike `prim_u8vector.scm`, this file does not use `check-same-behavior` against a `"~~lib/gambit/prim/bytevector#.scm"` namespace, because that file only stages the names in the `##unimplemented#` namespace while the working bindings come from the interpreter kernel — so there is no third lib-source namespace to compare against.

Two questions before I open a PR:

1. Is the $100 CAD bounty mentioned at the top of this issue still available?
2. If it is, would payment via PayPal to an account registered in mainland China work for you, or is another transfer method preferable?

Happy to open the PR right away if this direction looks right.

Thanks.

**Comment by feeley** [comment]
Procedures with no unit-tests can be listed with `make ut-coverage`. Currently the list contains 630 procedures so the 8 untested bytevector procedures are a small fraction of the total. The $100 CAD bounty is for the whole list, or at least a high proportion of it.

Are you doing this work with the help of an LLM? This has to be clearly stated if that is the case.

**Comment by tzh476** [comment]
Thanks for both answers — and yes, to be clear about the second question: **this is human + LLM collaboration.** I'm working with an LLM assistant that does the searching, writing and test-running; I set the direction, review the output and decide what gets posted. The coverage analysis and the candidate `bytevector.scm` in my previous comment were produced that way. I should have stated that upfront rather than waiting to be asked — sorry about that. If that makes the work unsuitable for this issue or for the bounty, I'd rather hear it now than waste your review time.

On the substance, thank you for pointing me at `make ut-coverage` — I didn't know about it and my hand-rolled analysis was a poor substitute. Running it here on `master` I get **663** procedures rather than the 630 you mention, so the list may have grown a little since you last looked (or my build differs). Either way your point stands: the 9 `bytevector` names are ~1.3% of it, not "a high proportion", so I understand the $100 CAD is not in play for what I've done so far.

Its composition, roughly by prefix, is:

```
char 55   string 31   thread 24   open 23   readtable 23
f32vector 13   f64vector 13   random 13   unbound 13   make 11
input 10   path 9   call 8   output 8   ...
```

which suggests it's tractable in coherent groups (the `char`/`string` block alone is ~13%) rather than as one heroic patch.

So, two things:

1. Would you like the `bytevector` file as a standalone PR anyway? It's written and passing (`FAILED 2 UNIT TESTS OUT OF 372` with it, `... OUT OF 371` without — same two pre-existing failures, `02-flonum/fladdmul.scm` and `03-number/lcm.scm`). It removes those 9 names from `ut-coverage`. Happy to open it purely as a contribution with no bounty attached, or to drop it if you'd rather not review LLM-assisted patches.
2. If a serious attempt at the whole list would be welcome, I'd propose working through it in reviewable batches by group (e.g. `char`, then `string`, then the `f32vector`/`f64vector` pair) rather than one large PR, so you can reject an approach early instead of after 663 procedures. But I won't start on that unless you say the LLM-assisted route is acceptable to you.

Either answer is fine — I'd just rather ask than assume.

**Comment by tzh476** [comment]
Thanks for pointing me at `make ut-coverage` — that changed how I'm approaching this, so here is a concrete status with numbers rather than intentions.

**On the count.** The list is now **583**, not 630, on a clean build of `ab6f255`. There is also a `ut-untested` target that extracts just the names, which is what I've been using.

**What I've actually finished: the `bytevector` group, all 12 of them.** My earlier draft only had 8 procedures; the list showed I'd missed four. So I completed the group:

| file | procedures |
|---|---|
| `07-vector/bytevector.scm` | `bytevector` `bytevector-append` `bytevector-copy` `bytevector-copy!` `bytevector-length` `bytevector-u8-ref` `bytevector-u8-set!` `bytevector?` |
| `09-io/bytevector_ports.scm` *(new)* | `make-bytevector` `open-input-bytevector` `open-output-bytevector` `get-output-bytevector` |

Measured effect on the coverage list: **583 → 571**, with exactly those 12 names disappearing (diffed name by name, not inferred from the total).

Test counts, with a baseline so the numbers mean something:

```
baseline (neither file)      373 tests, 3 failures
+ bytevector.scm             374 tests, 3 failures
+ bytevector_ports.scm       375 tests, 3 failures
```

The 3 failures are pre-existing on `ab6f255` and unrelated to this work — `02-flonum/fladdmul.scm` (`(fl+* -1. 0. -0.)` expects `-0.`, gets `0.`), `03-number/bitwise_eqv.scm` (`(bitwise-eqv)` expects `-1`, gets `0`), and `03-number/lcm.scm` (`(lcm 0 1.)` expects `0`, gets `0.`). Mentioning them only so it's clear I'm not adding failures; happy to leave them alone.

**One behaviour I'd rather ask about than assert.** `get-output-bytevector` returns the bytes accumulated *since the previous call* and resets the accumulator, so a second call after further writes yields only the new bytes rather than everything written so far. R7RS reads like the latter. But `get-output-string` behaves identically, so this looks like a deliberate, uniform Gambit design rather than a bytevector-specific bug — and I did **not** want to encode an R7RS reading that contradicts the implementation. The test currently pins the actual behaviour with a comment explaining why. Tell me if you'd prefer it asserted the other way, or left out.

**On the bounty, to be unambiguous:** I'm not claiming these 12 procedures earn the $100 CAD. You said the bounty is for the whole list or a high proportion of it, and 12 of 583 is 2.1%. I'm treating this as the first slice of a larger job.

**Remaining 571, grouped, so we can talk about order:**

```
string 28 · thread 24 · readtable 23 · open 21 · char 15 · random 13
unbound 13 · input 10 · make 9 · path 9 · call 8 · output 8 · tty 8 ...
```

Ranked by how well they suit unit tests, as I see it:

- `char` (~20) — pure predicates and conversions, no I/O, deterministic. Best next batch.
- `table` (7) — deterministic hash-table operations.
- `file` (7) — needs a temp directory, but the test tree already has helpers for that.
- `string` (28) — pure, though it overlaps SRFI-13 which has its own tests; worth checking for duplication first.
- `tcp` (6) / `address-info` (6) — need a listening socket or DNS; slower and more likely to be flaky.
- `thread` (24) — needs real synchronisation and is the easiest place to write tests that pass locally and hang in CI. I'd do this last, or not at all without your guidance.

**My question:** would you accept this as a series of small PRs, one group at a time, so you can reject an approach early rather than reviewing a huge diff? Or would you rather have one large PR at the end? I'd also welcome a hint on which groups you consider not worth unit-testing at all, so I don't spend effort where you'd reject it.

To restate the disclosure since it's relevant to reviewing volume: **this is human + LLM collaboration.** I'm working with an LLM assistant that does the searching, writing and test-running; I direct it, read the output and verify the numbers. Every figure above comes from a run on a clean tree, not an estimate. If that changes what you want from the contribution — including not wanting it — please say so and I'll stop rather than pile up PRs.

**Comment by gambiteer** [comment]
Thank you for your contribution.  I'm going to let Marc deal with this for now.

About the new test failures: These recently added/changed  tests accompany changes to the library, if you clone a recent GitHub head and rebuild, those failures should disappear.

**Comment by tzh476** [comment]
You were right, and I want to correct my own report rather than leave it standing.

The three failures were my build, not the library. Confirming the mechanism, since it wasn't a stale clone:

- My tree was at `ab6f255` — **0 commits behind** `origin/master`, i.e. literally the head whose message is "Fix zero-argument bitwise-eqv".
- But that commit touches `lib/_num.scm` and its test, **not** the pre-generated `lib/_num.c`. My build had compiled the older `.c`, so `gsi` still returned `0` for `(bitwise-eqv)` while the Scheme source said `-1`.
- Regenerating with `gsc-boot` (`touch lib/_num.scm && make _num.c`, then `make core`) changed the answer to `-1`.

Result after that rebuild:

```
PASSED ALL 375 UNIT TESTS
```

So `fladdmul`, `bitwise_eqv` and `lcm` all pass, and there were never any pre-existing failures to leave alone — I was reporting an artifact of building from checked-in C files. Apologies for the noise; I should have rebuilt before mentioning them.

Two things that did survive the rebuild:

1. **The coverage count is unchanged: 571**, identical name-for-name before and after the bootstrap. So that number and the 583 → 571 movement from the bytevector group are unaffected by this.
2. **375 tests pass with both bytevector files in place** — previously I could only say "same 3 failures as baseline", which was a weaker claim than it should have been.

The `get-output-bytevector` question and the batching question in my previous comment still stand whenever Marc has time. No rush.

**Comment by emanuelesaladini** [comment]
While checking the existing coverage collector on `ab6f255`, I reproduced two defects in `tests/unit-tests/#.scm` with `GAMBIT_COVERAGE=yes`:

- A `cond` test fails during expansion because both `eq?` checks receive one argument.
- A `case` with a variable key fails with `(cdr 'selector)`; other keys/bodies can also produce incorrect procedure lists.

A [focused fix and regression tests are prepared here](https://github.com/emanuelesaladini/gambit/commit/5886ee727cfd32779a503e38e36b93fc9d279a0f). The before-fix probes fail as described. On the rebuilt HEAD interpreter/compiler, all eight harness test files pass in interpreted and compiled modes, separately, with coverage on and off. Both full unit-suite runs (normal and coverage) pass 367 of 374 files; the same seven sandbox-dependent UDP/user-identity failures reproduce with the original collector. The new file includes ten collector assertions and two ordinary expression assertions.

Would you consider a separate CAD 10 bounty for this repair, payable through PayPal after acceptance? I understand that this does not fulfill the original CAD 100 scope. If the smaller scope is welcome, the focused PR is ready to open.

Disclosure: Codex authored the patch and tests and ran the checks on my behalf; this is AI-assisted work.

**Comment by danidin404** [comment]
@feeley I’m assessing the full CAD 100 unit-test bounty, rather than requesting payment for a small subset. This would be AI-assisted work: OpenAI Codex would author and run tests on my behalf, and that assistance would be disclosed in each submission.

Before undertaking broad coverage, could you confirm whether the bounty remains available to a new contributor, whether payment through PayPal is supported, and what coverage threshold would count as a “high proportion” of the currently untested procedures?

I would use the repository’s existing `make ut-coverage` collector, coordinate around open PRs, and report before/after coverage plus interpreted and compiled test results. I have not yet built the current source or produced new tests, and am not claiming any coverage improvement or award. Payment details would be provided privately only after acceptance.

**Comment by pythonyx135793** [comment]
Is the $100 CAD bounty still available for issue #756? If so, would you accept a staged PR that first adds an authoritative tested-procedure inventory and covers an agreed missing category, or must one PR add tests for every currently untested exported procedure? Payment by PayPal would work for me. I will wait for your scope confirmation before starting implementation.

**Comment by Da7em-T** [comment]
Opened a 2-file fix for the `GAMBIT_COVERAGE=yes` crash on `cond`/`case` expansion (`eq?` with one argument; `cdr` of the case key): https://github.com/gambit/gambit/pull/1043

Reproduced on Homebrew `gsi` 4.9.8 (exit 70 before, 0 after). Broader #756 unit-test slice remains #1042.

**Comment by Da7em-T** [comment]
Small #1009-style slice for still-untested R7RS names: https://github.com/gambit/gambit/pull/1044 (`make-parameter`, `case-lambda`, `define-record-type`, `make-promise`, `features`, `call/cc`). Separate from the coverage-collector fix in #1043 and the broader thread/list tests in #1042.

**Comment by Da7em-T** [comment]
One-file `pp` tests in the same shape as the merged #1009 print/println PR: https://github.com/gambit/gambit/pull/1045

**Comment by Da7em-T** [comment]
@feeley follow-up on the #756 coverage-collector crash: the two-file fix is https://github.com/gambit/gambit/pull/1043 (`tests/unit-tests/#.scm` walker + `00-test/test_coverage_cond_case.scm`).

Re-checked on Homebrew `gsi` 4.9.8: `gsi test_coverage_cond_case.scm` exits 0 with `GAMBIT_COVERAGE` unset and `yes`. Nested `cond`/`case` and one-clause `cond` are in the regression. If this slice is awardable under #756 / README donate-for-bounties, PayPal: **aa.631@hotmail.com**.
